src/compiler/checker.ts

```diff
@@ -12798,6 +12804,14 @@ namespace ts {
             return undefined;
         }
 
+        function getParameterTypeFromBody(parameter: ParameterDeclaration): Type {
+            const func = <FunctionLikeDeclaration>parameter.parent;
+            if (!func.body || isRestParameter(parameter)) return;
+
+            const types = checkAndAggregateParameterExpressionTypes(parameter);
+            return types ? getWidenedType(getIntersectionType(types)) : undefined;

```

Comment:
@HerringtonDarkholme I understand what you mean now; in the case you highlighted `a` should be inferred as `number | string` instead of `number & string`. `number & string` is an excessively conservative inference, but it is a sound one: `number & string` is assignable to `number | string`.

Some options are to bail and infer `any` for parameters used in branched code, keep the existing behavior and expect the user to manually supply typings when they realize `number & string` is unsatisfiable, or to actually analyze the flow graph and generate the appropriate union type.